### DISCUSSION TITLE
Usage of `extern "C" 'type'  NASM_function_name(int threadID); in carbon

### DISCUSSION THREAD

**Discussion by FCLC** [carbon-language/carbon-lang#1469] [General]
Hi guys/gals/NB-pals!

Currently going through the docs, trying to wrap my mind around if I should be using the same extern style as used in CPP to link to hand optimized SIMD/assembly functions. 

Currently you'd do something like 

`extern "C" 'type'  NASM_function_name(int threadID);`

where `NASM_function_name` is something along the lines of: 
```
NASM_function_name:
    push rbx
    push rdi
    push rsi
    push rbp
    push r11
    push r8
    push r9
    push r10
    push r12
```
/// math goes here
```
    dec r12
    jnz .LoopHead

    pop r12
    pop r10
    pop r9
    pop r8
    pop r11
    pop rbp
    pop rsi
    pop rdi
    pop rbx
    ret
```

in C++. 

you'd have NASM/yasm compile the *.asm and link via your makefile. 

what does that look like for carbon? what is the expected workflow in carbon?

**Comment by chandlerc** [comment]
Not sure what syntax we'll use for declaring external C functions when there isn't a C header to `import`, but yeah, we'll want to add some explicit syntax for that. Not sure its a high priority to design it until the rest of the interop is working, but then we'll probably work on some way of achieving this.

**Reply by FCLC** [reply]
Glad to hear! Project is interesting for certain, just hoping it doesn't go the way of https://killedbygoogle.com
